File: app/src/main/java/ru/tpu/russiantpu/main/dataAdapters/DocumentsDataAdapter.java
package ru.tpu.russiantpu.main.dataAdapters;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

import ru.tpu.russiantpu.R;
import ru.tpu.russiantpu.dto.DocumentDTO;

//DataAdapter для списка превью статей
public class DocumentsDataAdapter extends RecyclerView.Adapter<DocumentsDataAdapter.ViewHolder> {

    private LayoutInflater inflater;
    private List<DocumentDTO> items;
    private static ClickListener clickListener;

    public DocumentsDataAdapter(Context context, List<DocumentDTO> items) {
        this.items = items;
        this.inflater = LayoutInflater.from(context);
    }

    //создание объекта ViewHolder, хранящего данные по одному пункту DocumentDTO
    @NonNull
    @Override
    public DocumentsDataAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.item_document, parent, false);
        return new ViewHolder(view);
    }

    //привязка объекта ViewHolder к объекту пункта LinkItem
    @Override
    public void onBindViewHolder(DocumentsDataAdapter.ViewHolder holder, int position) {
        DocumentDTO item = items.get(position);

        String name = "\"" + item.getName() + "\""; //выводим название в кавычках
        holder.name.setText(name);
        holder.loadDate.setText(item.getLoadDate());
        holder.fileName.setText(item.getFileName());
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    //класс, хранящий элементы управления
    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {
        final TextView name;
        final TextView loadDate;
        final View downloadView;
        final TextView fileName;

        ViewHolder(View view){
            super(view);
            name = view.findViewById(R.id.doc_name);
            loadDate = view.findViewById(R.id.doc_date);
            downloadView = view.findViewById(R.id.doc_download);
            fileName = downloadView.findViewById(R.id.doc_file_name);
            //устанавливаем клик на контейнер элементов загрузки
            //(название файла и иконка)
            downloadView.setOnClickListener(this);
            downloadView.setOnLongClickListener(this);
        }

        @Override
        public void onClick(View v) {
            clickListener.onItemClick(getAdapterPosition(), v);
        }

        @Override
        public boolean onLongClick(View v) {
            clickListener.onItemLongClick(getAdapterPosition(), v);
            return false;
        }
    }

    public void setOnItemClickListener(ClickListener clickListener) {
        DocumentsDataAdapter.clickListener = clickListener;
    }
}

File: app/src/main/java/ru/tpu/russiantpu/main/dataAdapters/DocumentsDataAdapter.kt
package ru.tpu.russiantpu.main.dataAdapters

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import ru.tpu.russiantpu.R
import ru.tpu.russiantpu.dto.DocumentDTO

//DataAdapter для списка превью статей
class DocumentsDataAdapter(private val clickListener: (document: DocumentDTO) -> Unit)
    : RecyclerView.Adapter<DocumentsDataAdapter.ViewHolder>() {

    private var documents: List<DocumentDTO> = emptyList() // новые, не просмотренные документы
    private var newDocuments: MutableList<DocumentDTO> = mutableListOf()
    private var seenDocuments: MutableList<DocumentDTO> = mutableListOf() // уже скачанные, просмотренные доки
    private var newDocsOpen = true // дает знать, какие документы отображаются - новые или просмотренные

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DocumentsDataAdapter.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        return ViewHolder(inflater, parent)
    }

    override fun onBindViewHolder(holder: DocumentsDataAdapter.ViewHolder, position: Int) =
            holder.bind(documents[position])

    override fun getItemCount() = documents.size

    private fun initRecycler(list: List<DocumentDTO>) {
        if (documents !== list) {
            documents = list
            notifyDataSetChanged()
        }
    }

    /*
    * При обновлении recycler делим на два списка: непросмотренные и просмотренные документы
    * Если lastUseDate != null, значит документ был скачан хотя бы один раз, следовательно,
    * просмотрен.
    * */
    fun updateList(newDocuments: Collection<DocumentDTO>) {
        val (seenList, newList) = newDocuments.partition { it.lastUseDate != null }
        this.newDocuments = newList.toMutableList()
        this.seenDocuments = seenList.toMutableList()
        initRecycler(this.newDocuments)
    }

    fun showNewDocuments() {
        initRecycler(newDocuments)
        newDocsOpen = true
    }

    fun showSeenDocuments() {
        initRecycler(seenDocuments)
        newDocsOpen = false
    }

    inner class ViewHolder(inflater: LayoutInflater, parent: ViewGroup) :
            RecyclerView.ViewHolder(inflater.inflate(R.layout.item_document, parent, false)) {
        private val nameView: TextView = itemView.findViewById(R.id.doc_name)
        private val loadDateView: TextView = itemView.findViewById(R.id.doc_date)
        private val fileNameView: TextView = itemView.findViewById(R.id.doc_file_name)

        init {
            itemView.setOnClickListener {
                if (adapterPosition != RecyclerView.NO_POSITION) {
                    val clickedDoc = documents[adapterPosition]
                    clickListener(clickedDoc)
                    if (newDocsOpen) {
                        newDocuments.removeAt(adapterPosition)
                        documents = newDocuments
                        seenDocuments.add(clickedDoc)
                        notifyItemRemoved(adapterPosition)
                    }
                }
            }
        }

        fun bind(document: DocumentDTO) {
            val name = "\"" + document.name + "\"" //выводим название в кавычках
            nameView.text = name
            loadDateView.text = document.loadDate
            fileNameView.text = document.fileName
        }
    }
}